### PR TITLE
Reduce /<robot_name>/pose topic bandwidth

### LINKED ISSUE #261: CloudSim stops sending some topics
**Original report ([archived issue](https://osrf-migration.github.io/subt-gh-pages/#!/osrf/subt/issues/261)) by Martin Dlouhy (Bitbucket: [robotikacz](https://bitbucket.org/%7B27de76db-249c-44be-9eb8-91a3b17bf0d7%7D/)).**

----------------------------------------
The CloudSim or the bridge sometimes stop sending some topics. This issue was reported two months ago\(?\) in context of two running ROS nodes. Do you have some reference counters on bridge to check?

Here is the complete Python3 dump for robot A final run 10:

‌

```
0:00:01.874250 (0.0 0.0 0.0) []
Using times [10, 2400]
Artifacts before start: []
0:01:00.113899 (15.2 -3.6 0.0) [('acc', 1031), ('origin', 178), ('pose2d', 1029), ('rot', 1031), ('scan', 206), ('sim_time_sec', 21)]
0:02:00.041882 (37.3 -3.7 0.0) [('acc', 1109), ('pose2d', 1109), ('rot', 1109), ('scan', 222), ('sim_time_sec', 22)]
0:03:00.025751 (59.8 -3.6 0.0) [('acc', 1123), ('pose2d', 1123), ('rot', 1123), ('scan', 224), ('sim_time_sec', 23)]
0:04:00.039811 (74.7 3.7 0.0) [('acc', 1152), ('pose2d', 1153), ('rot', 1152), ('scan', 231), ('sim_time_sec', 23)]
0:05:00.087357 (74.7 27.0 0.0) [('acc', 1165), ('pose2d', 1164), ('rot', 1165), ('scan', 233), ('sim_time_sec', 23)]
0:06:00.032591 (58.9 33.6 0.0) [('acc', 1160), ('pose2d', 1161), ('rot', 1160), ('scan', 232), ('sim_time_sec', 23)]
0:07:00.091972 (37.8 28.3 0.0) [('acc', 1178), ('pose2d', 1177), ('rot', 1178), ('scan', 235), ('sim_time_sec', 24)]
0:08:00.037849 (36.8 9.7 -0.0) [('acc', 1167), ('pose2d', 1168), ('rot', 1167), ('scan', 234), ('sim_time_sec', 23)]
0:09:00.124006 (34.8 26.2 -0.0) [('acc', 1163), ('pose2d', 1162), ('rot', 1163), ('scan', 232), ('sim_time_sec', 24)]
0:10:00.099786 (51.8 36.4 -0.0) [('acc', 1143), ('pose2d', 1143), ('rot', 1143), ('scan', 229), ('sim_time_sec', 22)]
0:11:00.064612 (74.2 36.4 -0.0) [('acc', 1120), ('pose2d', 1120), ('rot', 1120), ('scan', 224), ('sim_time_sec', 23)]
0:12:00.037038 (74.8 58.0 -0.0) [('acc', 1147), ('pose2d', 1147), ('rot', 1147), ('scan', 229), ('sim_time_sec', 23)]
0:13:00.078897 (74.8 81.3 -0.0) [('acc', 1165), ('pose2d', 1165), ('rot', 1165), ('scan', 233), ('sim_time_sec', 23)]
0:14:00.084369 (77.4 67.2 -0.0) [('acc', 1170), ('pose2d', 1170), ('rot', 1170), ('scan', 234), ('sim_time_sec', 23)]
0:15:00.036357 (77.4 43.9 -0.0) [('acc', 1162), ('pose2d', 1163), ('rot', 1162), ('scan', 233), ('sim_time_sec', 24)]
0:16:00.101717 (92.1 36.2 -0.0) [('acc', 1157), ('pose2d', 1157), ('rot', 1157), ('scan', 231), ('sim_time_sec', 23)]
0:17:00.174657 (115.5 36.3 -0.0) [('acc', 1175), ('pose2d', 1174), ('rot', 1175), ('scan', 235), ('sim_time_sec', 23)]
0:18:00.071696 (139.0 36.3 -0.0) [('acc', 1174), ('pose2d', 1173), ('rot', 1174), ('scan', 235), ('sim_time_sec', 24)]
0:19:00.167036 (154.7 47.1 -0.0) [('acc', 1166), ('pose2d', 1167), ('rot', 1166), ('scan', 233), ('sim_time_sec', 23)]
0:20:00.112116 (154.7 70.5 -0.0) [('acc', 1153), ('pose2d', 1171), ('rot', 1153), ('scan', 234), ('sim_time_sec', 24)]
0:21:00.103524 (155.1 92.8 -0.0) [('pose2d', 1164), ('scan', 233), ('sim_time_sec', 23)]
0:22:00.040778 (155.6 115.8 -0.0) [('pose2d', 1152), ('scan', 230), ('sim_time_sec', 23)]
0:23:00.225396 (156.0 139.0 -0.0) [('pose2d', 1159), ('scan', 232), ('sim_time_sec', 23)]
0:24:00.096273 (156.5 162.0 -0.0) [('pose2d', 1151), ('scan', 231), ('sim_time_sec', 23)]
0:25:00.087319 (156.9 185.3 -0.0) [('pose2d', 1165), ('scan', 232), ('sim_time_sec', 23)]
0:26:00.017386 (157.4 208.3 -0.0) [('pose2d', 1147), ('scan', 230), ('sim_time_sec', 23)]
0:27:00.039352 (157.8 231.5 -0.0) [('pose2d', 1163), ('scan', 233), ('sim_time_sec', 24)]
0:28:00.071255 (158.3 255.2 -0.0) [('pose2d', 1185), ('scan', 236), ('sim_time_sec', 23)]
0:29:00.015956 (158.7 278.4 -0.0) [('pose2d', 1157), ('scan', 232), ('sim_time_sec', 23)]
0:30:00.025776 (159.2 301.8 -0.0) [('pose2d', 1173), ('scan', 234), ('sim_time_sec', 24)]
0:31:00.053610 (159.6 324.9 -0.0) [('pose2d', 1152), ('scan', 231), ('sim_time_sec', 23)]
0:32:00.156292 (160.1 347.5 -0.0) [('pose2d', 1130), ('scan', 226), ('sim_time_sec', 22)]
0:33:00.095004 (160.5 369.7 -0.0) [('pose2d', 1110), ('scan', 222), ('sim_time_sec', 23)]
0:34:00.020327 (160.9 391.6 -0.0) [('pose2d', 1096), ('scan', 219), ('sim_time_sec', 22)]
0:35:00.034189 (161.3 413.5 -0.0) [('pose2d', 1097), ('scan', 220), ('sim_time_sec', 22)]
0:36:00.085905 (161.8 435.2 -0.0) [('pose2d', 1085), ('scan', 216), ('sim_time_sec', 21)]
0:37:00.052622 (162.2 457.0 -0.0) [('pose2d', 1090), ('scan', 218), ('sim_time_sec', 22)]
0:38:00.038468 (162.6 478.4 -0.0) [('pose2d', 1070), ('scan', 215), ('sim_time_sec', 21)]
0:39:00.048765 (163.0 499.6 -0.0) [('pose2d', 1061), ('scan', 212), ('sim_time_sec', 22)]
0:40:00.166637 (163.4 520.7 -0.0) [('pose2d', 1055), ('scan', 211), ('sim_time_sec', 21)]
0:41:00.116588 (163.8 541.6 -0.0) [('pose2d', 1043), ('scan', 208), ('sim_time_sec', 21)]
0:42:00.005091 (164.2 562.5 -0.0) [('pose2d', 1047), ('scan', 210), ('sim_time_sec', 21)]
0:43:00.057229 (164.6 582.8 -0.0) [('pose2d', 1013), ('scan', 202), ('sim_time_sec', 20)]
0:44:00.086354 (165.0 603.6 -0.0) [('pose2d', 1040), ('scan', 208), ('sim_time_sec', 21)]
0:45:00.079366 (165.4 624.1 -0.0) [('pose2d', 1023), ('scan', 205), ('sim_time_sec', 20)]
0:46:00.094935 (165.8 645.1 -0.0) [('pose2d', 1053), ('scan', 210), ('sim_time_sec', 21)]
0:47:00.031867 (166.2 666.1 -0.0) [('pose2d', 1049), ('scan', 210), ('sim_time_sec', 21)]
0:48:00.119254 (166.6 687.0 -0.0) [('pose2d', 1046), ('scan', 209), ('sim_time_sec', 21)]
0:49:00.007596 (167.0 707.8 -0.0) [('pose2d', 1037), ('scan', 208), ('sim_time_sec', 21)]
0:50:00.054670 (167.5 728.9 -0.0) [('pose2d', 1058), ('scan', 211), ('sim_time_sec', 21)]
0:51:00.166188 (167.9 749.9 -0.0) [('pose2d', 1047), ('scan', 210), ('sim_time_sec', 21)]
0:52:00.006919 (168.3 770.5 -0.0) [('pose2d', 1030), ('scan', 206), ('sim_time_sec', 20)]
0:53:00.057601 (168.7 791.2 -0.0) [('pose2d', 1038), ('scan', 207), ('sim_time_sec', 21)]
0:54:00.026260 (169.1 811.5 -0.0) [('pose2d', 1049), ('scan', 210), ('sim_time_sec', 21)]
0:55:00.071333 (169.5 832.3 -0.0) [('pose2d', 1040), ('scan', 208), ('sim_time_sec', 21)]
0:56:00.042111 (169.9 853.2 -0.0) [('pose2d', 1046), ('scan', 209), ('sim_time_sec', 21)]
0:57:00.013416 (170.3 874.3 -0.0) [('pose2d', 1054), ('scan', 211), ('sim_time_sec', 21)]
0:58:00.119442 (170.7 895.7 -0.0) [('pose2d', 1070), ('scan', 214), ('sim_time_sec', 21)]
0:59:00.086211 (170.8 900.1 -0.0) [('pose2d', 220), ('scan', 216), ('sim_time_sec', 5)]
1:00:00.056517 (170.8 900.1 -0.0) [('scan', 214)]
1:01:00.139475 (170.8 900.1 -0.0) [('scan', 214)]
1:02:00.042254 (170.8 900.1 -0.0) [('scan', 213)]
1:03:00.064216 (170.8 900.1 -0.0) [('scan', 218)]
1:04:00.009180 (170.8 900.1 -0.0) [('scan', 219)]
1:05:00.098884 (170.8 900.1 -0.0) [('scan', 220)]
1:06:00.197615 (170.8 900.1 -0.0) [('scan', 225)]
1:07:00.028404 (170.8 900.1 -0.0) [('scan', 222)]
1:08:00.147307 (170.8 900.1 -0.0) [('scan', 220)]
1:09:00.194171 (170.8 900.1 -0.0) [('scan', 220)]
1:10:00.149403 (170.8 900.1 -0.0) [('scan', 219)]
1:11:00.247164 (170.8 900.1 -0.0) [('scan', 224)]
1:12:00.074530 (170.8 900.1 -0.0) [('scan', 224)]
1:13:00.038614 (170.8 900.1 -0.0) [('scan', 227)]
1:14:00.014625 (170.8 900.1 -0.0) [('scan', 223)]
1:15:00.061669 (170.8 900.1 -0.0) [('scan', 222)]
1:16:00.107241 (170.8 900.1 -0.0) [('scan', 219)]
1:17:00.246864 (170.8 900.1 -0.0) [('scan', 217)]
1:18:00.093432 (170.8 900.1 -0.0) [('scan', 217)]
1:19:00.180346 (170.8 900.1 -0.0) [('scan', 214)]
1:20:00.196161 (170.8 900.1 -0.0) [('scan', 215)]
1:21:00.276935 (170.8 900.1 -0.0) [('scan', 214)]
1:22:00.145624 (170.8 900.1 -0.0) [('scan', 213)]
1:23:00.139866 (170.8 900.1 -0.0) [('scan', 216)]
1:24:00.086690 (170.8 900.1 -0.0) [('scan', 215)]
1:25:00.112151 (170.8 900.1 -0.0) [('scan', 216)]
1:26:00.242472 (170.8 900.1 -0.0) [('scan', 217)]
1:27:00.093402 (170.8 900.1 -0.0) [('scan', 216)]
1:28:00.044074 (170.8 900.1 -0.0) [('scan', 215)]
1:29:00.065824 (170.8 900.1 -0.0) [('scan', 218)]
1:30:00.198382 (170.8 900.1 -0.0) [('scan', 218)]
1:31:00.064302 (170.8 900.1 -0.0) [('scan', 215)]
1:32:00.156794 (170.8 900.1 -0.0) [('scan', 213)]
1:33:00.116688 (170.8 900.1 -0.0) [('scan', 218)]
1:34:00.245220 (170.8 900.1 -0.0) [('scan', 216)]
1:35:00.120979 (170.8 900.1 -0.0) [('scan', 216)]
1:36:00.112766 (170.8 900.1 -0.0) [('scan', 216)]
1:37:00.024872 (170.8 900.1 -0.0) [('scan', 214)]
1:38:00.079595 (170.8 900.1 -0.0) [('scan', 215)]
1:39:00.269596 (170.8 900.1 -0.0) [('scan', 216)]
1:40:00.148613 (170.8 900.1 -0.0) [('scan', 211)]
1:41:00.064843 (170.8 900.1 -0.0) [('scan', 210)]
1:42:00.268545 (170.8 900.1 -0.0) [('scan', 216)]
1:43:00.125946 (170.8 900.1 -0.0) [('scan', 215)]
1:44:00.016775 (170.8 900.1 -0.0) [('scan', 208)]
1:45:00.108588 (170.8 900.1 -0.0) [('scan', 213)]
1:46:00.123156 (170.8 900.1 -0.0) [('scan', 213)]
1:47:00.198297 (170.8 900.1 -0.0) [('scan', 206)]
1:48:00.057000 (170.8 900.1 -0.0) [('scan', 207)]
1:49:00.021796 (170.8 900.1 -0.0) [('scan', 211)]
1:50:00.067160 (170.8 900.1 -0.0) [('scan', 212)]
1:51:00.106854 (170.8 900.1 -0.0) [('scan', 208)]
1:52:00.312889 (170.8 900.1 -0.0) [('scan', 205)]
1:53:00.120459 (170.8 900.1 -0.0) [('scan', 208)]
1:54:00.220170 (170.8 900.1 -0.0) [('scan', 207)]
1:55:00.211964 (170.8 900.1 -0.0) [('scan', 212)]
1:56:00.013743 (170.8 900.1 -0.0) [('scan', 206)]
1:57:00.140487 (170.8 900.1 -0.0) [('scan', 197)]
1:58:00.078270 (170.8 900.1 -0.0) [('scan', 203)]
1:59:00.022021 (170.8 900.1 -0.0) [('scan', 209)]
2:00:00.118374 (170.8 900.1 -0.0) [('scan', 212)]
2:01:00.015736 (170.8 900.1 -0.0) [('scan', 212)]
2:02:00.103999 (170.8 900.1 -0.0) [('scan', 210)]
2:03:00.195317 (170.8 900.1 -0.0) [('scan', 214)]
2:04:00.069948 (170.8 900.1 -0.0) [('scan', 212)]
2:05:00.144081 (170.8 900.1 -0.0) [('scan', 213)]
2:06:00.255350 (170.8 900.1 -0.0) [('scan', 210)]
2:07:00.038883 (170.8 900.1 -0.0) [('scan', 211)]
2:08:00.244349 (170.8 900.1 -0.0) [('scan', 208)]
2:09:00.220553 (170.8 900.1 -0.0) [('scan', 208)]
2:10:00.012854 (170.8 900.1 -0.0) [('scan', 210)]
2:11:00.182522 (170.8 900.1 -0.0) [('scan', 205)]
2:12:00.317423 (170.8 900.1 -0.0) [('scan', 211)]
2:13:00.329937 (170.8 900.1 -0.0) [('scan', 204)]
2:14:00.237473 (170.8 900.1 -0.0) [('scan', 211)]
2:15:00.102629 (170.8 900.1 -0.0) [('scan', 208)]
2:16:00.311819 (170.8 900.1 -0.0) [('scan', 202)]
2:17:00.064137 (170.8 900.1 -0.0) [('scan', 210)]
2:18:00.297227 (170.8 900.1 -0.0) [('scan', 211)]
2:19:00.011858 (170.8 900.1 -0.0) [('scan', 205)]
2:20:00.041207 (170.8 900.1 -0.0) [('scan', 208)]
2:21:00.306937 (170.8 900.1 -0.0) [('scan', 209)]
2:22:00.014030 (170.8 900.1 -0.0) [('scan', 207)]
2:23:00.057761 (170.8 900.1 -0.0) [('scan', 211)]
2:24:00.078889 (170.8 900.1 -0.0) [('scan', 211)]
2:25:00.271348 (170.8 900.1 -0.0) [('scan', 210)]
2:26:00.023680 (170.8 900.1 -0.0) [('scan', 207)]
2:27:00.206565 (170.8 900.1 -0.0) [('scan', 209)]
2:28:00.196093 (170.8 900.1 -0.0) [('scan', 214)]
2:29:00.190260 (170.8 900.1 -0.0) [('scan', 212)]
2:30:00.132740 (170.8 900.1 -0.0) [('scan', 213)]
2:31:00.065569 (170.8 900.1 -0.0) [('scan', 214)]
2:32:00.212321 (170.8 900.1 -0.0) [('scan', 213)]
2:33:00.196690 (170.8 900.1 -0.0) [('scan', 215)]
2:34:00.186527 (170.8 900.1 -0.0) [('scan', 215)]
2:35:00.349303 (170.8 900.1 -0.0) [('scan', 215)]
2:36:00.063184 (170.8 900.1 -0.0) [('scan', 210)]
2:37:00.077993 (170.8 900.1 -0.0) [('scan', 210)]
2:38:00.253989 (170.8 900.1 -0.0) [('scan', 213)]
2:39:00.205328 (170.8 900.1 -0.0) [('scan', 206)]
2:40:00.139102 (170.8 900.1 -0.0) [('scan', 204)]
2:41:00.172922 (170.8 900.1 -0.0) [('scan', 209)]
2:42:00.261287 (170.8 900.1 -0.0) [('scan', 210)]
2:43:00.193495 (170.8 900.1 -0.0) [('scan', 212)]
2:44:00.104197 (170.8 900.1 -0.0) [('scan', 209)]
2:45:00.166760 (170.8 900.1 -0.0) [('scan', 213)]
2:46:00.088262 (170.8 900.1 -0.0) [('scan', 208)]
2:47:00.194601 (170.8 900.1 -0.0) [('scan', 209)]
2:48:00.158934 (170.8 900.1 -0.0) [('scan', 211)]
2:49:00.034882 (170.8 900.1 -0.0) [('scan', 208)]
2:50:00.161464 (170.8 900.1 -0.0) [('scan', 202)]
2:51:00.113854 (170.8 900.1 -0.0) [('scan', 211)]
2:52:00.143288 (170.8 900.1 -0.0) [('scan', 211)]
2:53:00.111603 (170.8 900.1 -0.0) [('scan', 201)]
2:54:00.109043 (170.8 900.1 -0.0) [('scan', 208)]
```

‌

### PR DESCRIPTION
**Pull request by iche033**
Addresses issue #338, and potentially #389 and #261  

Depends on [ign-gazebo pull request 65](https://github.com/ignitionrobotics/ign-gazebo/pull/65) and [ros_ign pull request 67](https://github.com/osrf/ros_ign/pull/67)

New SDF params were added to the pose publisher system to help reduce bandwidth of the `/<robot_name>/pose` topic. The system now:
* subscribes to the new `/<robot_name>/pose_static` topic for static transforms (non-latched, published at 1Hz)
* changes the message type of the `/<robot_name>/pose[_static]` topics from `geometry_msgs/TransformStamped` to `tf2_messages/TFMessage` that consists a vector of transforms. This significantly reduces the number of messages that  need to be transmitted over the pose topic, e.g. for X1_C1, the publishing rate reduced from 2000Hz to 250Hz.

**Note:** The message type change is a breaking change but the `/<robot_name>/pose` topic is not an API that's publicly documented in https://github.com/osrf/subt/wiki/api. 

The changes address the critical / intermittent TF data loss issue for me. More tests are required to confirm whether or not this resolves #261 but with these changes I have not yet been able to reproduce the issue with message drops or delays on other topics. 

This PR also includes a heartbeat logger in BridgeLogger to help detect delays in messages.

Some ideas for testing:
* run subt and monitor `/tf` topic. Before the change, I see the the publishing rate slowly reduces and becomes intermittent
*  look for `Error: No messages received` or `Error: Missed message(s)` in ~/.ros. 
* Add the `pose` topic [here](https://github.com/osrf/subt/blob/master/subt_ros/src/BridgeLogger.cc#L118) to the list of topics monitored by BridgeLogger. Interestingly with this change, I start to see error messages about dropped messages in `~/.ros` within a few minutes of running subt.

### CODE REVIEW

File: subt_ros/src/BridgeLogger.cc [hunk 1]
```diff
@@ -211,6 +218,36 @@ void BridgeLogger::OnSensorMsg(const topic_tools::ShapeShifter::ConstPtr &_msg,
     return;
   }
 
+  // check heartbeat
```
Comment (nkoenig) [inline subt_ros/src/BridgeLogger.cc:221]: How about moving this heartbeat check to a separate callback that is triggered by a ROS timer? I think this code will work as long as some sensor data comes through. If all sensor data stops, then we won't get a heartbeat error.

Comment (iche033) [reply subt_ros/src/BridgeLogger.cc]: yep done. Moved heartbeat check into its own thread. 296de9e

**Comment by pauljurczak** [comment]
Is it included in 2020-04-21 release? I updated and I don't see these changes.

**Comment by nkoenig** [comment]
No, this is not in the 2020-04-21 release. Please refer to the [release notes on the wiki](https://github.com/osrf/subt/wiki/release_notes).

**Comment by iche033** [comment]
I can add a `Migration.md` guide for keeping track of changes like this one between circuits if that helps?

**Comment by nkoenig** [comment]
We have been putting release notes on the [wiki](https://github.com/osrf/subt/wiki/release_notes). I'll add a note there when this gets released.

**Comment by zwn** [comment]
Is there any time frame for when this will be released? I hope it will help with #415. I am not worried about migration since the original solution is not working reliably. If this indeed fixes the problems, I believe the teams will be more than happy to make any necessary adjustments.

**Comment by nkoenig** [comment]
> Is there any time frame for when this will be released? I hope it will help with #415. I am not worried about migration since the original solution is not working reliably. If this indeed fixes the problems, I believe the teams will be more than happy to make any necessary adjustments.

We are getting close. There are/were a few upstream releases that had to be deployed before this PR can be merged. Timeframe is likely this week.

**Pull request merged**